// @redocly/config/lib-esm/constants.js
export const DEFAULT_TEAM_CLAIM_NAME = 'https://redocly.com/sso/teams';
export var AuthProviderType;
(function (AuthProviderType) {
    AuthProviderType["OIDC"] = "OIDC";
    AuthProviderType["SAML2"] = "SAML2";
})(AuthProviderType || (AuthProviderType = {}));
export var ApigeeDevOnboardingIntegrationAuthType;
(function (ApigeeDevOnboardingIntegrationAuthType) {
    ApigeeDevOnboardingIntegrationAuthType["SERVICE_ACCOUNT"] = "SERVICE_ACCOUNT";
    ApigeeDevOnboardingIntegrationAuthType["OAUTH2"] = "OAUTH2";
})(ApigeeDevOnboardingIntegrationAuthType || (ApigeeDevOnboardingIntegrationAuthType = {}));
export const REDOCLY_TEAMS_RBAC = 'redocly::teams-rbac';
export const REDOCLY_ROUTE_RBAC = 'redocly::route-rbac';
export var LayoutVariant;
(function (LayoutVariant) {
    LayoutVariant["STACKED"] = "stacked";
    LayoutVariant["THREE_PANEL"] = "three-panel";
})(LayoutVariant || (LayoutVariant = {}));
export const ENTITY_RELATION_TYPES = [
    'partOf',
    'hasParts',
    'creates',
    'createdBy',
    'owns',
    'ownedBy',
    'implements',
    'implementedBy',
    'dependsOn',
    'dependencyOf',
    'uses',
    'usedBy',
    'produces',
    'consumes',
    'linksTo',
    'supersedes',
    'supersededBy',
    'compatibleWith',
    'extends',
    'extendedBy',
    'relatesTo',
    'hasMember',
    'memberOf',
    'triggers',
    'triggeredBy',
    'returns',
    'returnedBy',
];
export const NODE_TYPE_NAMES = {
    USER_ENTITY: 'UserEntity',
    USER_ENTITY_METADATA: 'UserEntityMetadata',
    API_DESCRIPTION_ENTITY: 'ApiDescriptionEntity',
    API_DESCRIPTION_METADATA: 'ApiDescriptionEntityMetadata',
    API_OPERATION_ENTITY: 'ApiOperationEntity',
    API_OPERATION_METADATA: 'ApiOperationEntityMetadata',
    DATA_SCHEMA_ENTITY: 'DataSchemaEntity',
    DATA_SCHEMA_METADATA: 'DataSchemaEntityMetadata',
    SERVICE_ENTITY: 'ServiceEntity',
    DOMAIN_ENTITY: 'DomainEntity',
    TEAM_ENTITY: 'TeamEntity',
    ENTITY: 'Entity',
    ENTITY_METADATA: 'EntityMetadata',
    ENTITY_LINK_LIST: 'EntityLinkList',
    ENTITY_LINK: 'EntityLink',
    ENTITY_RELATION: 'EntityRelation',
    ENTITY_RELATION_LIST: 'EntityRelationList',
    ENTITY_CONTACT: 'EntityContact',
    SLACK_CONTACT: 'EntitySlackContact',
    SLACK_CHANNEL: 'EntitySlackChannel',
};
