// @kubb/plugin-client/dist/clients/fetch.cjs
Object.defineProperty(exports, '__esModule', { value: true });
const require_chunk = require('../chunk-C1_xRkKa.cjs');

//#region src/clients/fetch.ts
let _config = {};
const getConfig = () => _config;
const setConfig = (config) => {
	_config = config;
	return getConfig();
};
const client = async (paramsConfig) => {
	const normalizedParams = new URLSearchParams();
	const globalConfig = getConfig();
	const config = {
		...globalConfig,
		...paramsConfig,
		headers: {
			...Array.isArray(globalConfig.headers) ? Object.fromEntries(globalConfig.headers) : globalConfig.headers,
			...Array.isArray(paramsConfig.headers) ? Object.fromEntries(paramsConfig.headers) : paramsConfig.headers
		}
	};
	Object.entries(config.params || {}).forEach(([key, value]) => {
		if (value !== void 0) normalizedParams.append(key, value === null ? "null" : value.toString());
	});
	let targetUrl = [config.baseURL, config.url].filter(Boolean).join("");
	if (config.params) targetUrl += `?${normalizedParams}`;
	const response = await fetch(targetUrl, {
		credentials: config.credentials || "same-origin",
		method: config.method?.toUpperCase(),
		body: config.data instanceof FormData ? config.data : JSON.stringify(config.data),
		signal: config.signal,
		headers: config.headers
	});
	return {
		data: [
			204,
			205,
			304
		].includes(response.status) || !response.body ? {} : await response.json(),
		status: response.status,
		statusText: response.statusText,
		headers: response.headers
	};
};
client.getConfig = getConfig;
client.setConfig = setConfig;
var fetch_default = client;

//#endregion
exports.client = client;
exports.default = fetch_default;
exports.getConfig = getConfig;
exports.setConfig = setConfig;
