// @redocly/config/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NODE_TYPE_NAMES = exports.ENTITY_RELATION_TYPES = exports.LayoutVariant = exports.REDOCLY_ROUTE_RBAC = exports.REDOCLY_TEAMS_RBAC = exports.ApigeeDevOnboardingIntegrationAuthType = exports.AuthProviderType = exports.DEFAULT_TEAM_CLAIM_NAME = void 0;
exports.DEFAULT_TEAM_CLAIM_NAME = 'https://redocly.com/sso/teams';
var AuthProviderType;
(function (AuthProviderType) {
    AuthProviderType["OIDC"] = "OIDC";
    AuthProviderType["SAML2"] = "SAML2";
})(AuthProviderType || (exports.AuthProviderType = AuthProviderType = {}));
var ApigeeDevOnboardingIntegrationAuthType;
(function (ApigeeDevOnboardingIntegrationAuthType) {
    ApigeeDevOnboardingIntegrationAuthType["SERVICE_ACCOUNT"] = "SERVICE_ACCOUNT";
    ApigeeDevOnboardingIntegrationAuthType["OAUTH2"] = "OAUTH2";
})(ApigeeDevOnboardingIntegrationAuthType || (exports.ApigeeDevOnboardingIntegrationAuthType = ApigeeDevOnboardingIntegrationAuthType = {}));
exports.REDOCLY_TEAMS_RBAC = 'redocly::teams-rbac';
exports.REDOCLY_ROUTE_RBAC = 'redocly::route-rbac';
var LayoutVariant;
(function (LayoutVariant) {
    LayoutVariant["STACKED"] = "stacked";
    LayoutVariant["THREE_PANEL"] = "three-panel";
})(LayoutVariant || (exports.LayoutVariant = LayoutVariant = {}));
exports.ENTITY_RELATION_TYPES = [
    'partOf',
    'hasParts',
    'creates',
    'createdBy',
    'owns',
    'ownedBy',
    'implements',
    'implementedBy',
    'dependsOn',
    'dependencyOf',
    'uses',
    'usedBy',
    'produces',
    'consumes',
    'linksTo',
    'supersedes',
    'supersededBy',
    'compatibleWith',
    'extends',
    'extendedBy',
    'relatesTo',
    'hasMember',
    'memberOf',
    'triggers',
    'triggeredBy',
    'returns',
    'returnedBy',
];
exports.NODE_TYPE_NAMES = {
    USER_ENTITY: 'UserEntity',
    USER_ENTITY_METADATA: 'UserEntityMetadata',
    API_DESCRIPTION_ENTITY: 'ApiDescriptionEntity',
    API_DESCRIPTION_METADATA: 'ApiDescriptionEntityMetadata',
    API_OPERATION_ENTITY: 'ApiOperationEntity',
    API_OPERATION_METADATA: 'ApiOperationEntityMetadata',
    DATA_SCHEMA_ENTITY: 'DataSchemaEntity',
    DATA_SCHEMA_METADATA: 'DataSchemaEntityMetadata',
    SERVICE_ENTITY: 'ServiceEntity',
    DOMAIN_ENTITY: 'DomainEntity',
    TEAM_ENTITY: 'TeamEntity',
    ENTITY: 'Entity',
    ENTITY_METADATA: 'EntityMetadata',
    ENTITY_LINK_LIST: 'EntityLinkList',
    ENTITY_LINK: 'EntityLink',
    ENTITY_RELATION: 'EntityRelation',
    ENTITY_RELATION_LIST: 'EntityRelationList',
    ENTITY_CONTACT: 'EntityContact',
    SLACK_CONTACT: 'EntitySlackContact',
    SLACK_CHANNEL: 'EntitySlackChannel',
};
